content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,317 @@
+        }
+    ]
+}
+```
+{{% notes type="info" %}}
+Let's break this down a bit: 
+
+* `type: node` tells VS Code that we are using the Node runtime
+* `port: 9292` tells VS Code to attach the debugger to port 9292
+* `request: attach` tells VS Code that we want to attach to to the debugger at launch
+* `preLaunchTask: pulumi-debug` tells VS Code to run the task we configured in `tasks.json` before we connect the debugger. This will run our `pulumi-debug.sh` script and then attach to the Node.js runtime 
+* `continueOnAttach: true` tells VS Code to start running our code as soon as we attach to the waiting Node.js runtime (until we reach our first breakpoint that is, where it will pause and return control to the debugger client, aka VS Code)
+* `restart: {...}` configures VS Code to wait 5 seconds and then retry connecting, up to 5 times. This is here in case it takes a while for the Node.js runtime to start running. It's unlikely, but possible!
+* `skipFiles: {...}` sets up some filters of files *not* to try to debug. This includes everything that is part of node's internals, 3rdparty node modules, and other libraries. That way we don't end up deep in someone else's code unexpectedly and can just focus on our Pulumi program.
+* `name` sets a human-readable name which shows up in the drop down box in VS Code when you choose which [launch configuration][vs-code-launch-configs] to run. 
+{{% /notes %}}
+
+Finally, we need to make one more edit to the `pulumi-debug.sh` file we created. Here will will pass the necessary options to the Node.js runtime to set it to run in debug mode. 
+
+Add the following line to the beginning of the file:
+
+```shell {hl_lines=[3]}
+#!/bin/bash
+
+export NODE_OPTIONS='--inspect-brk=127.0.0.1:9292'
+
+pulumi login
+pulumi stack select dev
+pulumi up -f
+```
+{{% notes type="info" %}}
+The `NODE_OPTIONS` environment variable passes options to the Node.js V8 runtime. In this case, we are passing `--inspect-brk` which tells the Node.js runtime to run in debug mode, which will cause it to pause before running the code, until a debugging client connects to `localhost` on port `9292`.
+{{% /notes %}}
+
+Ok, now we should be ready to run our Pulumi program from VS Code using the `F5` key! Let's give it a try. You should see Pulumi running inside of the "Terminal" tab. You will also see output indicating that it is waiting for a debugger to attach: 
+
+```
+Debugger listening on ws://127.0.0.1:9292/########-####-####-####-############
+For help, see: https://nodejs.org/en/docs/inspector
+Debugger attached.
+Waiting for the debugger to disconnect...
+``` 
+
+### Step 4: Setting a debugging breakpoint 
+
+Now we can try [setting a breakpoint][vs-code-breakpoints] in our code. Navigate to a line within the code and hit the `F9` key. You'll see a small red dot appear next to the line indicating the breakpoint. You can then run the program again using `F5`. When the execution of the code reaches the line it will pause there.
```

Comment:
One thing to note is that the debugging happens in a different terminal session/window, so if your Pulumi program relies on environment variables (e.g., your AWS creds), you'll need some way to pass them to that session. For example, I kept hitting this error (even after ensuring my AWS creds were current in the embedded terminal) until I realized the debugger wasn't actually using that terminal:

```plain
Diagnostics:
  pulumi:providers:aws (default_6_37_1):
    error: pulumi:providers:aws resource 'default_6_37_1' has a problem: unable to validate AWS credentials.
    Details: validating provider credentials: retrieving caller identity from STS: operation error STS: GetCallerIdentity, https response error StatusCode: 403, RequestID: 8e37f328-dfa8-4234-8234-5cca32d6ea48, api error ExpiredToken: The security token included in the request is expired
```

There's an `env` property you can set in `launch.json` to help with this -- e.g., if you use a named `AWS_PROFILE` (as you probably wouldn't want to hack your actual `AWS_*` creds into that file). For me, since I use AWS SSO through ESC, adding this line to the shell script before `pulumi up` did the trick:

```bash 
eval "$(pulumi env open pulumi/dev-sandbox --format shell)"
```

Reply:
Since this configuration will vary so much by the user's local config (e.g. Windows vs Mac, how they have creds set-up for AWS, etc), I've chosen to just leave this part unexplained rather than to try to figure out a way to explain it in a way that is exhaustive enough to cover all the likely scenarios.